### PR TITLE
View Images by Tag

### LINKED ISSUE #9: Images by Tag
#### As a user I want to view all images associated with a tag.

__Story__:

Now that we have added the ability for images to be associated with tags, let
us provide the ability for users to filter images by a given tag. Nana, who
loves fast planes, will be delighted to see only images labeled with SR-71.

__Acceptance criteria__:
- [ ] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [ ] The tag name, not its ID, is used to perform the filtering.

__Discussion topics__:
- [ ] Where in the code should we implement this functionality?
- [ ] What happens if a user tries to manually filter by a nonexistent tag?

__Dependencies__:
- Image Tags

### PR DESCRIPTION
**Pull request by awright022**
Closes #9 

Add ability to click on a tag name and pull up all images for that tag.  

I implemented this on the rails controller level; javascript based solutions seemed too simplistic, and likely could have caused longer term headaches.

### CODE REVIEW

File: app/controllers/home_controller.rb [hunk 1]
```diff
@@ -1,5 +1,7 @@
 class HomeController < ApplicationController
   def index
-    @images = Image.all.reverse
+    @images = Image.eager_load(:tags).all
+    params[:tag] && @images = @images.tagged_with(params[:tag])
+    @images = @images.reverse
```
Comment (shiboying) [inline app/controllers/home_controller.rb]: This part is clean an nice! Just one thing, it is more user-friendly to add a flash message to tell the user where there is no image founded with this tag.

File: app/views/images/_image.html.erb [hunk 2]
```diff
@@ -1,8 +1,8 @@
 <div class="img-wrap" id="img-wrap-<%= image.id %>">
   <%= image_tag image.url, class: 'homepage-images' %>
   <p class="card-text">
-    <% image.tag_list.each do |tag| %>
-        <span class="image-tag"><%= tag %></span>
+    <% image.tags.each do |tag| %>
+        <a href="/?tag=<%= tag %>"><span class="image-tag"><%= tag %></span></a>
```
Comment (shiboying) [inline app/views/images/_image.html.erb]: Instead of using the raw html style, you could use `link_to` method which we use a lot. And you could do something like: `<%= link_to tag, images_path(tag: tag), class: 'image-tag' %>`.

File: test/controllers/home_controller_test.rb [hunk 3]
```diff
@@ -27,10 +27,31 @@ class HomeControllerTest < ActionDispatch::IntegrationTest
...
     assert_response :ok
-    assert_select '.image-tag' do |elements|
-      assert_equal 2, elements.count
+    assert_select '.image-tag', count: 2 do |elements|
       assert_equal elements[0].text, image.tag_list.first
       assert_equal elements[1].text, image.tag_list[1]
     end
   end
+
+  test 'should get index with image by tag' do
```
Comment (shiboying) [inline test/controllers/home_controller_test.rb]: Besides this, you could also add a test for no use tag, for example if a user modify the url and try to access some random tag themselves.

Also, you could add a flow test for this task: `show images by tag`.

Comment (awright022) [inline test/controllers/home_controller_test.rb]: Hi!  Thanks for the feedback.  I'm adding in the test for a tag with empty results now, and the alert message you referenced for the empty result case.

But I do have a question about the flow test you referenced; do you think it will contribute value?  Because this is all implemented on the server-side level, and that is all tested by the controller/integration tests, I don't really think that a flow test would offer any different test coverage than the current test suite.

I think I'd prefer to omit that flow test rather than duplicating the effort and weighing down the tests with extra selenium tests for no added benefit.

Comment (shiboying) [inline test/controllers/home_controller_test.rb]: From test coverage perspective, I agree with you. And what I suggested to add a flow test is from practice point of view. If you feel comfortable writing selenium tests and you are familiar with page object, you can skip this step.
